fix(github): Fix github issue assignment feature

imkhateeb opened the pull request:
![Screenshot 2025-04-07 175705](https://github.com/user-attachments/assets/59507929-7b01-480f-8272-331385772b74)
![Screenshot 2025-04-07 175723](https://github.com/user-attachments/assets/48eba3e8-90cb-429f-97d8-1da54f91445b)

lalitindoria commented:
Please add details about how did you find that admin:org scope is required for assignment?

imkhateeb commented:
`admin:org` scope for OAuth apps in github gives complete control over organization, for example to get all members (including private members) of an organization.

**How this help in assignment of an issue to a user?**
When we give prompt to assign an issue to an `xyz` user(See #134 description), it first get all members and then assign it to the user who matches with name/username/id `xyz`

batflarrow commented:
> `admin:org` scope for OAuth apps in github gives complete control over organization, for example to get all members (including private members) of an organization.
> 
> **How this help in assignment of an issue to a user?** When we give prompt to assign an issue to an `xyz` user(See #134 description), it first get all members and then assign it to the user who matches with name/username/id `xyz`

Have you tested if it was working before the change?
Also, can you share some documentation that states that the search user API requires this scope?
Was the API failing earlier when the scope was not added? I don't find any such requirement of scope here https://docs.github.com/en/rest/search/search?apiVersion=2022-11-28#search-users

imkhateeb commented:
It is not clearly written anywhere, but ChatGPT suggested this:

> ​To access a GitHub organization's private member information, the `admin:org` OAuth scope is indeed required. This scope grants full administrative access to the organization and its teams, projects, and memberships. ​

> The read:org scope, on the other hand, provides read-only access to organization membership, organization projects, and team membership. However, it does not grant access to private members' details. ​

Before adding this scope `admin:org` I was getting error while assigning a org repo issue to a member, but assignment was working for personal repository. After using this assignment feature is working fine for both personal and org repo.

I will add example in the next comment.

batflarrow commented:
> It is not clearly written anywhere, but ChatGPT suggested this:
> 
> > ​To access a GitHub organization's private member information, the `admin:org` OAuth scope is indeed required. This scope grants full administrative access to the organization and its teams, projects, and memberships. ​
> 
> > The read:org scope, on the other hand, provides read-only access to organization membership, organization projects, and team membership. However, it does not grant access to private members' details. ​
> 
> Before adding this scope `admin:org` I was getting error while assigning a org repo issue to a member, but assignment was working for personal repository. After using this assignment feature is working fine for both personal and org repo.
> 
> I will add example in the next comment.

Yes, Please mention the exact error that you got before and after this change.

imkhateeb commented:
@batflarrow You are right, we don't need `admin:org` scope to get users of an organization. And for me the issue is getting assigned without any change.

batflarrow commented:
Closing this PR. Please re-open when you've identified the actual issue.

imkhateeb commented:
I am reopening this PR because I think I find out the problem. After thorough testing I will add results in next comment.

imkhateeb commented:
### The problem here was:
- When we were providing a valid and complete username(e.g., imkhateeb) then assignment of issue was working fine.
- But, when we were providing partial username(e.g., khateeb)
    - Sometimes it work well because it fetches all the user from orgs and proceed with the best match
    - Sometimes it didn't work because it assumes `khateeb` as complete and valid username and proceed with that.

### How I fixed this issue:
- I introduced one validation inside `addAssigneeToIssue` and `removeAssigneeFromIssue` function which will fetch all org users and match it with the provided username(full or half) everytime.

### Result:
![Screenshot 2025-04-17 115335](https://github.com/user-attachments/assets/de29b103-d42b-4e53-b744-bf93c307d162)
![Screenshot 2025-04-17 115315](https://github.com/user-attachments/assets/be1a5dbe-65e6-4784-bd8e-3e0787dac209)

batflarrow commented on agent-packages/packages/github/src/index.ts:
```diff
@@ -71,6 +71,25 @@ export class GitHubService implements BaseService<GitHubConfig> {
   }
 
+  private async resolveUsername(orgName: string, username: string): Promise<string> {
+    const userSearch = await this.getUsers(orgName);
```
Use search users API

imkhateeb replied:
I have tried users search API with org filter, but it is not working.

batflarrow reviewed (commented):
Let's fix the prompt such that if in the context we need a user, search_user tool is called by the LLM.

batflarrow commented:
> Let's fix the prompt such that if in the context we need a user, search_user tool is called by the LLM.

Instead of modifying the tool selection prompt, you update the description of the assignee property to say that use search_user tool to find out the username to assign the issue to. Ask ChatGPT to come up with a better version of this description

batflarrow commented on agent-packages/packages/github/src/tools.ts:
```diff
@@ -158,7 +158,11 @@ export async function createGitHubToolsExport(config: GitHubConfig): Promise<Too
+        assignee: z
+          .string()
+          .describe(
+            'The GitHub username of the assignee (e.g. "imkhateeb" not just "khateeb"). Provide the exact GitHub username by using the "get_organization_users" tool first then assign the user to the issue. PRs and Issues are interchangeable terms in GitHub'
```
```suggestion
            'The GitHub username of the assignee (e.g. "imkhateeb" not just "khateeb"). Provide the exact GitHub username by using the "get_organization_users" tool first then assign the user to the issue.'
```
